src/approaches/grammar_search_invention_approach.py

```diff
@@ -396,6 +397,220 @@ def _create_grammar(grammar_name: str, dataset: Dataset,
     raise NotImplementedError(f"Unknown grammar name: {grammar_name}.")
 
 
+################################################################################
+#                            Heuristic Functions                               #
+################################################################################
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _PredicateSearchHeuristic:
+    """A heuristic for guiding search over predicate sets.
+    """
+    initial_predicates: Set[Predicate]  # predicates given by the environment
+    atom_dataset: List[GroundAtomTrajectory]  # data with all candidates applied
+    candidates: Dict[Predicate, float]  # candidate predicates to costs
+
+    def evaluate(self, predicates: FrozenSet[Predicate]) -> float:
+        """Get the heuristic value for the set of predicates.
+
+        Lower is better.
+        """
+        raise NotImplementedError("Override me!")
+
+    def _get_predicate_penalty(self, predicates: FrozenSet[Predicate]) -> float:
+        """Get a heuristic penalty based on the predicate complexities.
+        """
+        pred_complexity = sum(self.candidates[p] for p in predicates)
+        return CFG.grammar_search_pred_complexity_weight * pred_complexity
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _OperatorLearningBasedHeuristic(_PredicateSearchHeuristic):
+    """A heuristic that learns operators given the set of predicates.
+    """
+
+    def evaluate(self, predicates: FrozenSet[Predicate]) -> float:
+        print("Evaluating predicates:", predicates)
+        pruned_atom_data = utils.prune_ground_atom_dataset(
+            self.atom_dataset, predicates | self.initial_predicates)
+        segments = [seg for traj in pruned_atom_data
+                    for seg in segment_trajectory(traj)]
+        strips_ops, partitions = learn_strips_operators(segments,
+                                                        verbose=False)
+        option_specs = [p.option_spec for p in partitions]
+        op_heuristic = self._evaluate(predicates, pruned_atom_data, segments,
+                                      strips_ops, option_specs)
+        pred_penalty = self._get_predicate_penalty(predicates)
+        op_penalty = self._get_operator_penalty(strips_ops)
+        total_score = op_heuristic + pred_penalty + op_penalty
+        print("Total score:", total_score)
+        return total_score
+
+    def _evaluate(self, predicates: FrozenSet[Predicate],
```

Comment:
Great refactor here! This name feels too similar to the parent, while it's really doing a more specific thing. Rename to `def _evaluate_with_operators`?